lib/clerk/authenticate_context.rb

```diff
@@ -0,0 +1,189 @@
+require "ostruct"
+require "forwardable"
+require "base64"
+
+module Clerk
+    ##
+    # This class represents a parameter object used to contain all request and configuration
+    # information required by the middleware to resolve the current request state.
+    # link: https://refactoring.guru/introduce-parameter-object
+    class AuthenticateContext
+        extend Forwardable
+
+        ##
+        # Expose the url of the request that this parameter object was created from as a URI object.
+        attr_reader :clerk_url
+
+        ##
+        # Expose properties that does not require validations or complex logic to retrieve
+        # values by delegating them to the cookies or headers variables.
+        def_delegators :@cookies, :session_token_in_cookie, :client_uat
+        def_delegators :@headers, :session_token_in_header, :sec_fetch_dest
+
+        ##
+        # Creates a new parameter object using Rack::Request and Clerk::Config objects.
+        def initialize(request, config)
+            @clerk_url = URI.parse(request.url)
+            @config = config
+
+            @cookies = OpenStruct.new({
+                session_token_in_cookie: request.cookies[SESSION_COOKIE],
+                client_uat: request.cookies[CLIENT_UAT_COOKIE],
+                handshake_token: request.cookies[HANDSHAKE_COOKIE],
+                dev_browser: request.cookies[DEV_BROWSER_COOKIE]
+            })
+
+            header_token = request.env[AUTHORIZATION_HEADER] || ''
+            @headers = OpenStruct.new({
+                session_token_in_header: header_token.gsub(/bearer/i, '').strip,
+                sec_fetch_dest: request.env[SEC_FETCH_DEST_HEADER],
+                accept: request.env[ACCEPT_HEADER] || "",
+                origin: request.env[ORIGIN_HEADER] || "",
+                host: request.host,
+                port: request.port
+            })
+        end
+
+        ##
+        # The following properties are part of the props supported in all the AuthenticateContext
+        # objects across all of our SDKs (eg JS, Go)
+        def secret_key
+            @config.api_key || ""
```

Comment:
:wink: Same for the rest.

```suggestion
            @config.api_key.to_s
```